processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java

```diff
@@ -1121,59 +1200,237 @@ public boolean isLiteral() {
     }
 
     /**
-     * Steps through the declaredType in order to find a match for this typevar Type. It allignes with
+     * Steps through the declaredType in order to find a match for this typeVar Type. It aligns with
      * the provided parameterized type where this typeVar type is used.
      *
-     * @param declaredType the type
-     * @param parameterizedType the parameterized type
+     * For example:
+     * this: T
+     * declaredType: JAXBElement<String>
+     * parameterizedType: JAXBElement<T>
+     * result: String
+     *
+     *
+     * this: T, T[] or ? extends T,
+     * declaredType: E.g. Callable<? extends T>
+     * parameterizedType: Callable<BigDecimal>
+     * return: BigDecimal
+     *
+     * @param declared the type
+     * @param parameterized the parameterized type
      *
-     * @return the matching declared type.
+     * @return - the same type when this is not a type var in the broadest sense (T, T[], or ? extends T)
+     *         - the matching parameter in the parameterized type when this is a type var when found
+     *         - null in all other cases
      */
-    public Type resolveTypeVarToType(Type declaredType, Type parameterizedType) {
-        if ( isTypeVar() ) {
-            TypeVarMatcher typeVarMatcher = new TypeVarMatcher( typeUtils, this );
-            return typeVarMatcher.visit( parameterizedType.getTypeMirror(), declaredType );
+    public ResolvedPair resolveParameterToType(Type declared, Type parameterized) {
```

Comment:
We need it.. The problem is that you don't necessarily line up a typevar `T`, but it might just as well be `T[]` or `? extends T`. You don't know that when you call this `resolveParameterToType` with the type parameters you've declared on the method.

Look also at the call site, in MethodMatcher, line 254. So, you always return what you matched (lined up), and not per-se the typevar itself.

Reply:
I'll get the code locally and have a look at it. It is not a deal breaker anyways :)

Reply:
You are right, not sure what I saw back then :)